src/types/capsule.rs

```diff
@@ -203,37 +232,159 @@ impl PyCapsule {
     ///
     /// It must be known that this capsule is valid and its pointer is to an item of type `T`.
     pub unsafe fn reference<T>(&self) -> &T {
-        &*(self.pointer() as *const T)
+        self.as_borrowed().reference()
     }
 
     /// Gets the raw `c_void` pointer to the value in this capsule.
     ///
     /// Returns null if this capsule is not valid.
     pub fn pointer(&self) -> *mut c_void {
+        self.as_borrowed().pointer()
+    }
+
+    /// Checks if this is a valid capsule.
+    ///
+    /// Returns true if the stored `pointer()` is non-null.
+    pub fn is_valid(&self) -> bool {
+        self.as_borrowed().is_valid()
+    }
+
+    /// Retrieves the name of this capsule, if set.
+    ///
+    /// Returns an error if this capsule is not valid.
+    pub fn name(&self) -> PyResult<Option<&CStr>> {
+        self.as_borrowed().name()
+    }
+
+    fn name_ptr_ignore_error(slf: &Bound<'_, Self>) -> *const c_char {
```

Comment:
I think keeping this as an inherent method is a bit detrimental to readability. I would suggest making it a free function within this module and maybe putting it next to `ensure_no_error`.